crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -738,9 +752,55 @@ where
                 is_star,
                 range: _,
             }) => {
+                // Save the state prior to visiting any of the `try` block.
+                //
+                // Potentially none of the `try` block could have been executed prior to executing
+                // the `except` block(s), so we will merge this state with all of the intermediate
+                // states during the `try` block before visiting the `except` blocks.
+                let pre_try_block_state = self.flow_snapshot();
+                self.try_block_contexts().push_try_block();
+
+                // Visit the `try` block!
+                //
+                // If `visiting_try_block == true`, this informs `.add_definition()` to record the state
+                // in `self.try_block_definition_states` after each new definition is recorded.
                 self.visit_body(body);
 
-                for except_handler in handlers {
+                // Save the state immediately *after* visiting the `try` block
+                // but *before* we prepare for visiting the `except` block(s).
+                //
+                // We will revert to this state prior to visiting the the `else` block,
+                // as there necessarily must have been 0 `except` blocks executed
+                // if we hit the `else` block.
+                let post_try_block_state = self.flow_snapshot();
+
+                // Prepare for visiting the `except` block(s)
+
+                let visiting_try_block_states = self
+                    .try_block_contexts()
+                    .pop_try_block()
+                    .expect("A `try` block should have been pushed to the stack prior to visiting the `body` field");
+                if let Some(parent_try_block) =
+                    self.try_block_contexts().borrow_mut().current_try_block()
+                {
+                    parent_try_block.record_visiting_nested_try_stmt();
```

Comment:
The method name here seems misleading to me because it is called **after** we visit the `try` body. It seems it's about visiting the except handlers, finaly , and or else blocks of a `try`. 

Related: What's the reason that we don't record definitions in a nested try? E.g. what if

```python
try:
    try:
        x =  10:
        maybe_panic()
    finally:
        y = panic()
finally:
    y
```

shouldn't the use of `y` be an error? 

but 

```python
try:
    try:
        x =  10:
        maybe_panic()
    finally:
        y = 10
finally:
    y
```

should be okay?

Reply:
(Cases to consider for tests here are e.g. `x = 1; x = 2` inside an `except`, `else`, or `finally` block in a nested try statement -- the `x = 1` possibility should be visible to the outer try block.

Reply:
> (And such an exception would not cause us to still enter an inner `finally` block, which only applies to the inner `try` body; it would jump straight to being handled by the outer `try` block.)

What do you mean? The inner `finally` block is indeed always executed, even if the exception is caught by the outer `try` block:

```pycon
>>> try:
...     try:
...         raise TypeError()
...     except NameError:
...         print(1)
...     finally:
...         print(2)
... except TypeError:
...     print(3)
... finally:
...     print(4)
... 
2
3
4
```

Reply:
What I mean is that if an exception is raised inside an except/else/finally block in the nested try (not inside the try body), in _that_ case the inner finally is not executed, the exception unwinds directly to the outer try statement handling.

Reply:
I guess the one thing to consider still is the case where an exception is raised within the `finally` block itself? I.e. is it important that we register those intermediate states within the nested `finally` block with the outer try statement?

Reply:
I still can't find any cases where the inner `finally` is not executed, though it is of course true that if an exception is raised in an `except` handler then we might switch to the inner `finally` block before the except handler has finished, and indeed if an exception is raised in the inner `finally` block then that `finally` block might not run to completion. I might not handle _that_ properly right now...

```pycon
>>> try:
...     try:
...         raise TypeError
...     except TypeError:
...         raise RuntimeError
...         print(1)
...     finally:
...         print(2)
... except RuntimeError:
...     print(3)
... finally:
...     print(4)
... 
2
3
4
```